solr/core/src/java/org/apache/solr/pkg/PackageAwareSolrClassLoader.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.pkg;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.BiFunction;
+
+import org.apache.solr.core.PluginInfo;
+import org.apache.solr.core.SolrClassLoader;
+import org.apache.solr.core.SolrCore;
+import org.apache.solr.core.SolrResourceLoader;
+import org.apache.solr.util.plugin.SolrCoreAware;
+
+import static java.util.Collections.singletonMap;
+
+/**
+ * This class implements a SolrClassLoader that can  identify the correct packages
+ * and load classes from that. This also listens to any changes to the relevant packages and
+ * invoke a callback if anything is modified
+ */
+public class PackageAwareSolrClassLoader implements SolrClassLoader {
+  final SolrCore core;
+  final SolrResourceLoader loader;
+  private Map<String, PackageAPI.PkgVersion> classNameVsPkg = new HashMap<>();
+
+  private final List<PackageListeners.Listener> listeners = new ArrayList<>();
+  private final Runnable reloadRunnable;
+
+
+  public PackageAwareSolrClassLoader(SolrCore core, SolrResourceLoader loader, Runnable runnable) {
```

Comment:
What's the runnable for?

Reply:
It's to reload your set of plugins. In the case of schema it's a `refreshSchema()` , If you wish to reload core or something else , you could do it here